api/src/database/migrations/20240328A-permissions-policies.ts

```diff
@@ -0,0 +1,197 @@
+import type { Knex } from 'knex';
+import { randomUUID } from 'node:crypto';
+import { PUBLIC_POLICY_ID } from '@directus/constants';
+import { processChunk } from '@directus/utils';
+
+/**
+ * The public role used to be `null`, we gotta create a single new policy for the permissions
+ * previously attached to the public role (marked through `role = null`).
+ * And a new public role that is marked by a different, arbitrary UUID.
+ */
+
+export async function up(knex: Knex) {
+	/////////////////////////////////////////////////////////////////////////////////////////////////
+	// Create new policies table that mirrors previous Roles
+
+	await knex.schema.createTable('directus_policies', (table) => {
+		table.uuid('id').primary();
+		table.string('name', 100).notNullable();
+		table.text('description');
+		table.text('ip_access');
+		table.boolean('enforce_tfa').defaultTo(false).notNullable();
+		table.boolean('admin_access').defaultTo(false).notNullable();
+		table.boolean('app_access').defaultTo(false).notNullable();
+	});
+
+	/////////////////////////////////////////////////////////////////////////////////////////////////
+	// Copy over all existing roles into new policies
+
+	const roles = await knex
+		.select('id', 'name', 'description', 'ip_access', 'enforce_tfa', 'admin_access', 'app_access')
+		.from('directus_roles');
+
+	if (roles.length > 0) {
+		await processChunk(roles, 100, async (chunk) => {
+			await knex('directus_policies').insert(chunk);
+		});
+	}
+
+	await knex
+		.insert({
+			id: PUBLIC_POLICY_ID,
+			name: '$t:public_label',
+			description: '$t:public_description',
+			app_access: false,
+		})
+		.into('directus_policies');
+
+	// Change the admin policy description to $t:admin_policy_description
+	await knex('directus_policies')
+		.update({
+			description: '$t:admin_policy_description',
+		})
+		.where({ description: '$t:admin_description' });
+
+	/////////////////////////////////////////////////////////////////////////////////////////////////
+	// Remove access control + add nesting to roles
+
+	await knex.schema.alterTable('directus_roles', (table) => {
+		table.dropColumn('ip_access');
+		table.dropColumn('enforce_tfa');
+		table.dropColumn('admin_access');
+		table.dropColumn('app_access');
+
+		table.uuid('parent').references('directus_roles.id');
```

Comment:
There's the foreign key error when a parent role is deleted.
Reminder to rename the migration file to a later date.

```suggestion
		table.uuid('parent').references('directus_roles.id').onDelete('SET NULL');
```

Reply:
> Reminder to rename the migration file to a later date.

Already noted in the top level todos 🫡

Reply:
Let's rename now 😃 I don't expect another migration to sweep in.

Reply:
I didn't do it for now since that would screw up our testing DBs. But I need to touch it later today anyways in order to add the `icon` field. So I might as well do it then.

Reply:
The sooner the better, so we test against the final state 👍